File: ios/Sources/DeckStore.swift
import Foundation
import Observation

/// The decks the app knows about.
///
/// The list lives here rather than in `@State` on the view, for the reason
/// Chapter 5 gives: a rule about *what may be added* is logic, and logic on a
/// view can only be checked by tapping. Here it is seven tests.
@Observable
final class DeckStore {
    private(set) var decks: [Deck]

    init(decks: [Deck] = DeckStore.sample) {
        self.decks = decks
    }

    static let sample: [Deck] = [
        Deck(name: "Spanish — Verbs", totalCount: 240),
        Deck(name: "Spanish — Food", totalCount: 86),
        Deck(name: "Kanji N5", totalCount: 103),
    ]

    /// Whether a proposed name can be added.
    ///
    /// Trimmed, non-empty, and not already taken. The comparison is
    /// case-insensitive because "kanji n5" and "Kanji N5" are the same deck to
    /// everyone except a string.
    func canAdd(_ name: String) -> Bool {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return false }
        return !decks.contains { $0.name.lowercased() == trimmed.lowercased() }
    }

    /// Adds a deck, if the name is acceptable.
    ///
    /// - Returns: whether it was added, so a caller can tell the difference
    ///   between "done" and "silently ignored".
    @discardableResult
    func add(named name: String) -> Bool {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard canAdd(trimmed) else { return false }
        decks.append(Deck(name: trimmed, totalCount: 0))
        return true
    }
}

File: android/app/src/main/kotlin/com/ashkanghaderi/flashcards/DeckStore.kt
package com.ashkanghaderi.flashcards

/**
 * The decks the app knows about, and the rule about what may be added.
 *
 * Port of DeckStore.swift. The rule lives here rather than in the Activity for
 * exactly the reason Chapter 8 gives - on a view it could only be checked by
 * typing into an emulator.
 */
class DeckStore(
    decks: List<Deck> = sample,
    private val storage: DeckStorage? = null,
) {
    var decks: List<Deck> = decks
        private set

    var loadError: String? = null
        private set

    companion object {
        val sample: List<Deck> = Deck.sample

        /** Loads from storage, keeping the failure rather than hiding it. */
        fun loading(storage: DeckStorage): DeckStore {
            val store = DeckStore(emptyList(), storage)
            try {
                store.decks = storage.load()
            } catch (error: StorageException) {
                store.loadError = error.message
            }
            return store
        }
    }

    fun canAdd(name: String): Boolean {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return false
        return decks.none { it.name.equals(trimmed, ignoreCase = true) }
    }

    fun add(name: String): Boolean {
        val trimmed = name.trim()
        if (!canAdd(trimmed)) return false
        decks = decks + Deck(trimmed, 0)
        persist()
        return true
    }

    private fun persist() {
        val storage = storage ?: return
        if (loadError != null) return          // Chapter 9: never write over it
        runCatching { storage.save(decks) }
    }
}
